### PR TITLE
 Fix issue of Multiple API Calls 

### LINKED ISSUE #112: APIs are calling more than once
### Description:

Whenever any user performs any functionality in EmbeddedChat, the getMessage API is called more than once which eventually results in status code 429 due to multiple requests. This issue might act as a blocker in implementing more features in the future.

### Steps to reproduce:

1. Send any message or react on any message
2. Do this multiple times

### Expected behavior:

The getMessage API should be called only once instead of calling itself multiple times.

### Actual behavior:

The API is calling multiple times due to which after some operations we get status code 429.

### PR DESCRIPTION
**Pull request by VipinDevelops**
# Fixing the issue of getMessage API being called multiple times

## Description
Whenever any user performs any functionality in EmbeddedChat, the getMessage API is called more than once which eventually results in status code 429 due to multiple requests. This issue might act as a blocker in implementing more features in the future. This PR aims to fix this issue by modifying the real-time function in api.js to only call the API once when a user sends a message or reacts to a message.

## Acceptance Criteria fulfillment
The getMessage API is only called once when a user sends a message or reacts to a message
Multiple requests are prevented, avoiding the status code 429 error
No regressions are introduced

## Changes Made
Modified the real-time function in api.js to use a Promise that resolves the first time the onMessage event is triggered, ensuring the getMessage API is only called once.

Fixes 
#112  and #132  Task 1 
- [x]  Multiple API requests being made 

## Video 
[Screencast from 2023-02-18 11-31-49.webm](https://user-images.githubusercontent.com/99081689/219844303-e359e1e6-1c85-4fc6-848e-0aa9fc9f58c3.webm)

### CODE REVIEW

**Comment by VipinDevelops** [comment]
Hey, @sidmohanty11 I Fixed the issue of getMessage API being called multiple times resulting in status code 429.

I've thoroughly tested the solution and it's working as expected. Could you please take a look at the PR and let me know if you have any feedback or suggestions for improvement? I would greatly appreciate your review.

Thank you!!

File: src/lib/api.js [hunk 1]
```diff
@@ -229,15 +229,25 @@ export default class RocketChatInstance {
...
       await this.rcClient.resume({ token: Cookies.get(RC_USER_TOKEN_COOKIE) });
-      await this.rcClient.subscribe('stream-room-messages', this.rid);
-      await this.rcClient.onMessage((data) => {
+      subscriptionId = await this.rcClient.subscribe(
+        'stream-room-messages',
+        this.rid
+      );
+
+      const messageHandler = (data) => {
+        this.rcClient.unsubscribe(subscriptionId);
```
Comment (sidmohanty11) [inline src/lib/api.js:242]: If we are unsubscribing it, are we maintaining the realtime chat functionality?

File: src/lib/api.js [hunk 2]
```diff
@@ -229,15 +229,25 @@ export default class RocketChatInstance {
...
         callback(data);
-      });
+      };
+
+      await this.rcClient.onMessage(messageHandler);
```
Comment (sidmohanty11) [inline src/lib/api.js:246]: is `onMessage` room specific or called everytime?

**Review by sidmohanty11 — CHANGES_REQUESTED** [review]
Thanks a lot @VipinDevelops for the PR! I had some questions

File: src/lib/api.js (hunk 1 above)
Comment (VipinDevelops) [reply src/lib/api.js:242]: Yes i showed that in preview also real-time communication is working perfectly fine

File: src/lib/api.js (hunk 2 above)
Comment (VipinDevelops) [reply src/lib/api.js:246]: it is called for all messages received by the client and  The streamDataHandler function is also not room-specific, but it filters events based on the room ID before calling the callback.

**Comment by sidmohanty11** [comment]
@VipinDevelops this is not the right way in which we should handle it. I think @abhinavkrin and @vishakha-aggarwal mentioned a better way and we should proceed with that.

**Pull request closed without merging**